### PR TITLE
Vit

### PR DESCRIPTION
**Pull request by refor53**

### CODE REVIEW

File: ViT/model/README_code.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        self.patch_size = patch_size
+        super().__init__()
+        self.projection = nn.Sequential(
+            # using a conv layer instead of a linear one -> performance gains
+            nn.Conv2d(in_channels, emb_size, kernel_size=patch_size, stride=patch_size),
+            Rearrange('b e (h) (w) -> b (h w) e'),
+        )
+        self.cls_token = nn.Parameter(torch.randn(1,1, emb_size))
+        self.positions = nn.Parameter(torch.randn((img_size // patch_size) **2 + 1, emb_size))
+
```
Comment (refor53) [inline ViT/model/README_code.py:12]: 224x224 이미지를 가정해서 함수 __init__ 선언, PatchEmbedding에 필요한 변수 선언
1. linear projection
2. cls_token 추가
3. positions 추가 - 총 패치 개수만큼 만들어서 크기 일치

File: ViT/model/README_code.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    def forward(self, x: Tensor) -> Tensor:
+        b, _, _, _ = x.shape
+        x = self.projection(x)
+        cls_tokens = repeat(self.cls_token, '() n e -> b n e', b=b)
```
Comment (Raymondgwangryeol) [inline ViT/model/README_code.py:16]: `einops.repeat(tensor, 'input_dims -> output_dims', variable=values) `
cls_token을 b크기에 맞춰 반복.    
cls_tokens size => (batch_size, 1, embedding size)

File: ViT/model/README_code.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        # prepend the cls token to the input
+        x = torch.cat([cls_tokens, x], dim=1)
```
Comment (Alexaqua7) [inline ViT/model/README_code.py:18]: 이미지 패치 앞에 CLS Token을 삽입하여 Sequence Length (dim=1) 방향으로 concate

File: ViT/model/README_code.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        return x
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, emb_size: int = 768, num_heads: int = 8, dropout: float = 0):
+        super().__init__()
+        self.emb_size = emb_size
+        self.num_heads = num_heads
+        # fuse the queries, keys and values in one matrix
+        self.qkv = nn.Linear(emb_size, emb_size * 3)
```
Comment (Alexaqua7) [inline ViT/model/README_code.py:30]: query, key, value를 한 개의 nn.Linear로 하여 처리. query, key, value 각각 emb_size의 sequence length를 가짐

File: ViT/model/README_code [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            nn.Conv2d(in_channels, emb_size, kernel_size=patch_size, stride=patch_size),
+            Rearrange('b e (h) (w) -> b (h w) e'),
+        )
+        self.cls_token = nn.Parameter(torch.randn(1,1, emb_size))
+        self.positions = nn.Parameter(torch.randn((img_size // patch_size) **2 + 1, emb_size))
+
+    def forward(self, x: Tensor) -> Tensor:
+        b, _, _, _ = x.shape
+        x = self.projection(x)
+        cls_tokens = repeat(self.cls_token, '() n e -> b n e', b=b)
```
Comment (zaqquum) [inline ViT/model/README_code]: - [1,1, 768] -> [b= 1, 1, 768] 로 변환 (b=1임으로 변화 없음)
- `einops.repeat(input, pattern)` :  torch.repeat과 유사

File: ViT/model/README_code.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        cls_tokens = repeat(self.cls_token, '() n e -> b n e', b=b)
+        # prepend the cls token to the input
+        x = torch.cat([cls_tokens, x], dim=1)
+        # add position embedding
+        x += self.positions
+        return x
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, emb_size: int = 768, num_heads: int = 8, dropout: float = 0):
```
Comment (Alexaqua7) [inline ViT/model/README_code.py:25]: emb_size = patch_size ** 2 * num_channel

File: ViT/model/README_code.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+class ResidualAdd(nn.Module):
+    def __init__(self, fn):
+        super().__init__()
+        self.fn = fn
+        
+    def forward(self, x, **kwargs):
+        res = x
+        x = self.fn(x, **kwargs)
+        x += res
+        return x
```
Comment (ken001111) [inline ViT/model/README_code.py:63]: 이렇게 하는거군요

File: ViT/model/README_code.py (hunk 2 above)
Comment (mnmsgit) [reply ViT/model/README_code.py:16]: self.cls_token = nn.Parameter(torch.randn(1,1, emb_size)) 이렇게 초기화 되었던 코드가
batch 개수만큼 반복하여 생성하여 (bath_size,1,emb_size) 가 되는것이구나 :)

File: ViT/model/README_code.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        self.emb_size = emb_size
+        self.num_heads = num_heads
+        # fuse the queries, keys and values in one matrix
+        self.qkv = nn.Linear(emb_size, emb_size * 3)
+        self.att_drop = nn.Dropout(dropout)
+        self.projection = nn.Linear(emb_size, emb_size)
+        
+    def forward(self, x : Tensor, mask: Tensor = None) -> Tensor:
+        # split keys, queries and values in num_heads
+        qkv = rearrange(self.qkv(x), "b n (h d qkv) -> (qkv) b h n d", h=self.num_heads, qkv=3)
```
Comment (mnmsgit) [inline ViT/model/README_code.py:36]: self.qkv(x)의 shape 은 (batch_size, seq_len, h * d * qkv) 형태 인데 이를 (qkv, batch_size, h, seq_len, d)로 reshpae하는 효과, 후의 queries, keys, values 에 대한 indexing을 편하게 하기 위해 사용한 것 같습니다.

File: ViT/model/README_code.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        queries, keys, values = qkv[0], qkv[1], qkv[2]
+        # sum up over the last axis
+        energy = torch.einsum('bhqd, bhkd -> bhqk', queries, keys) # batch, num_heads, query_len, key_len
+        if mask is not None:
+            fill_value = torch.finfo(torch.float32).min
+            energy.mask_fill(~mask, fill_value)
+
```
Comment (mnmsgit) [inline ViT/model/README_code.py:43]: Transformer에서 봤던 attention 계산을 그대로 수행하되 dropout을 추가하였습니다.

File: ViT/model/README_code.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                nn.Dropout(drop_p)
+            )
+            ))
+
+class TransformerEncoder(nn.Sequential):
+    def __init__(self, depth: int = 12, **kwargs):
+        super().__init__(*[TransformerEncoderBlock(**kwargs) for _ in range(depth)])
+
+
+class ClassificationHead(nn.Sequential):
```
Comment (mnmsgit) [inline ViT/model/README_code.py:102]: 여기서는 Classification을 하기위한 MLP head에서 CLS토큰만을 이용하지 않고 평균 pooling을 이용하여 classification을 진행 한 것 같습니다

File: ViT/model/README_code.py (hunk 1 above)
Comment (zaqquum) [reply ViT/model/README_code.py:12]: 성능 향상을 위해 Liner layer -> Conv2d layer 사용

output : torch.Size([1, 196, 768])
nn.Conv2d(in_channels, out_channels, kernel_size, stride, padding)

File: ViT/model/README_code.py (hunk 2 above)
Comment (zaqquum) [reply ViT/model/README_code.py:16]: [1,1, 768] -> [b= 1, 1, 768] 로 변환 (b=1임으로 변화 없음)
einops.repeat(input, pattern) : torch.repeat과 유사
